src/main/java/io/anserini/util/mapper/CountDocumentMapper.java

```diff
@@ -0,0 +1,114 @@
+package io.anserini.util.mapper;
+
+import io.anserini.collection.SourceDocument;
+import io.anserini.util.MapCollections;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang3.time.DurationFormatUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.File;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class CountDocumentMapper extends DocumentMapper {
+
+  private static final Logger LOG = LogManager.getLogger(CountDocumentMapper.class);
+
+  public final static class Counters {
+    /**
+     * Counter for successfully indexed documents.
+     */
+    public AtomicLong indexed = new AtomicLong();
+
+    /**
+     * Counter for empty documents that are not indexed. Empty documents are not necessary errors;
+     * it could be the case, for example, that a document is comprised solely of stopwords.
+     */
+    public AtomicLong empty = new AtomicLong();
+
+    /**
+     * Counter for unindexable documents. These are cases where {@link SourceDocument#indexable()}
+     * returns false.
+     */
+    public AtomicLong unindexable = new AtomicLong();
+
+    /**
+     * Counter for skipped documents. These are cases documents are skipped as part of normal
+     * processing logic, e.g., using a whitelist, not indexing retweets or deleted tweets.
+     */
+    public AtomicLong skipped = new AtomicLong();
+
+    /**
+     * Counter for unexpected errors.
+     */
+    public AtomicLong errors = new AtomicLong();
+  }
+
+  private final Counters counters;
+  private final Set whitelistDocids;
+
+  public CountDocumentMapper(MapCollections.Args args) throws Exception {
+    super(args);
+    this.counters = new CountDocumentMapper.Counters();
+
+    if (args.whitelist != null) {
+      List<String> lines = FileUtils.readLines(new File(args.whitelist), "utf-8");
+      this.whitelistDocids = new HashSet<>(lines);
+    } else {
+      this.whitelistDocids = null;
+    }
+  }
+
+  public void incrementErrors() {
+    counters.errors.incrementAndGet();
+  }
+
+  public void incrementSkipped() {
+    counters.skipped.incrementAndGet();
+  }
+
+  public void incrementUnindexable() {
+    counters.unindexable.incrementAndGet();
+  }
+
+  public void incrementEmpty() {
+    counters.empty.incrementAndGet();
+  }
+
+  public void incrementIndexed() {
+    counters.indexed.incrementAndGet();
+  }
+
+  public void incrementIndexedBy(int numIndexed) {
+    counters.indexed.addAndGet(numIndexed);
+  }
+
+  public boolean process(SourceDocument d) {
+    if (!d.indexable()) {
+      incrementUnindexable();
```

Comment:
Is there a reason why you dispatch to a method here as opposed to just incrementing the counter directly?

And furthermore, does it make sense for incrementFoo methods to be public?

Reply:
1) Huhhh, actualy, no reason.

2) I made them public because there I have to call them in multiple places inside `MapCollections`. There is no way to put all the counter updates inside `process()` and keep the signature of current `process()` method unchanged. So I used something like:
```
if (mapper.isCountDocumentMapper()) {
  ((CountDocumentMapper)mapper).incrementSkipped();
}
```
inside `MapCollections`.

Reply:
I see. This is a bit janky but I can't think of anything better right now... so this is fine, and we'll let this evolve as we find more use cases for mappers.